### PR TITLE
[QHC-57] Implement platform.execute_qprogram() for quantum machines

### PR DESCRIPTION
**Pull request by fedonman**
Closes QHC-57, QHC-109, QHC-110, QHC-157, QHC-178, QHC-177, QHC-189

### CODE REVIEW

File: docs/releases/changelog-dev.md [hunk 1]
```diff
@@ -2,6 +2,20 @@
 
 ### New features since last release
 
+- Allow execution of `QProgram` through `platform.execute_qprogram` method for Quantum Machines hardware.
+  [#648](https://github.com/qilimanjaro-tech/qililab/pull/648)
+
+### Improvements
+
+- Added `bus_mapping` parameter in `QbloxCompiler.compile` method to allow changing the bus names of the compiled output.
```
Comment (visagim) [inline docs/releases/changelog-dev.md:10]: I don't know if the standard is to do what you did here - putting different Improvements / Features belonging to the same PR separately - or putting them together (just wondering if there's a standard at all)

File: src/qililab/instruments/qblox/qblox_qrm.py [hunk 2]
```diff
@@ -139,12 +139,14 @@ def _get_qprogram_acquisitions(self, acquisitions: list[str]) -> list[QbloxQProg
...
                     )
+                    self.device.store_scope_acquisition(sequencer=sequencer.identifier, name=acquisition)
                     raw_measurement_data = self.device.get_acquisitions(sequencer=sequencer.identifier)[acquisition][
                         "acquisition"
                     ]
                     measurement_result = QbloxQProgramMeasurementResult(raw_measurement_data=raw_measurement_data)
                     results.append(measurement_result)
-
+        for sequencer in self.awg_sequencers:
+            self.device.delete_acquisition_data(sequencer=sequencer.identifier, all=True)
```
Comment (visagim) [inline src/qililab/instruments/qblox/qblox_qrm.py:149]:
```suggestion
        
                self.device.delete_acquisition_data(sequencer=sequencer.identifier, all=True)
```
can we just do this inside the `sequencer in self.awg_sequencers` loop above?

File: src/qililab/instruments/quantum_machines/quantum_machines_cluster.py [hunk 3]
```diff
@@ -282,34 +285,85 @@ def reset(self):
...
+        Raises:
+            NotImplementedError: Raised if not connected to Quantum Machines
+            ParameterNotFound: Raised if parameter does not exist
+        """
+        # TODO: This is a temporary solution. Implement again when QM API gets updated.
+        if not self._is_connected_to_qm:
+            raise NotImplementedError("You should be connected to Quantum Machines in order to change a parameter.")
+        if parameter == Parameter.LO_FREQUENCY:
+            self._qm.octave.set_lo_frequency(element=bus, lo_frequency=float(value))
+            if "RF_inputs" in self._config["elements"][bus]:
```
Comment (visagim) [inline src/qililab/instruments/quantum_machines/quantum_machines_cluster.py:324]: if `parameter == Parameter.LO_FREQUENCY` but `("RF_inputs" in self._config["elements"][bus]) == False` then we will raise a `ParameterNotFound` error (but technically the parameter does exist). Should this happen?

Comment (visagim) [reply src/qililab/instruments/quantum_machines/quantum_machines_cluster.py:324]: Same for other parameters below

File: src/qililab/instruments/yokogawa/gs200.py [hunk 4]
```diff
@@ -198,6 +198,7 @@ def setup(self, parameter: Parameter, value: float | str | bool, channel_id: int
     @Instrument.CheckDeviceInitialized
     def initial_setup(self):
         """Performs an initial setup."""
+        self.device.off()
         self.source_mode = self.settings.source_mode
```
Comment (visagim) [inline src/qililab/instruments/yokogawa/gs200.py:202]: how is this related to the pr description `Implement platform.execute_qprogram() for quantum machines`?

File: src/qililab/platform/platform.py [hunk 5]
```diff
@@ -579,15 +590,48 @@ def execute_qprogram(self, qprogram: QProgram) -> dict[str, list[Result]]:
...
-        qblox_compiler = QProgramQbloxCompiler()
-        sequences = qblox_compiler.compile(qprogram=qprogram)
+        qblox_compiler = QbloxCompiler()
+        sequences = qblox_compiler.compile(qprogram=qprogram, bus_mapping=bus_mapping)
         buses = {bus_alias: self._get_bus_by_alias(alias=bus_alias) for bus_alias in sequences}
 
+        if debug:
+            with open("debug.py", "w", encoding="utf-8") as sourceFile:
+                for bus_alias in sequences:
+                    print(bus_alias, file=sourceFile)
```
Comment (visagim) [inline src/qililab/platform/platform.py:625]: add more descriptive comments

File: src/qililab/qprogram/qblox_compiler.py [hunk 6]
```diff
@@ -103,10 +103,11 @@ def __init__(self):
         }
 
         self._qprogram: QProgram
+        self._bus_mapping: dict[str, str] | None
         self._buses: dict[str, BusCompilationInfo]
         self._sync_counter: int
 
-    def compile(self, qprogram: QProgram) -> dict[str, QPy.Sequence]:
+    def compile(self, qprogram: QProgram, bus_mapping: dict[str, str] | None = None) -> dict[str, QPy.Sequence]:
```
Comment (visagim) [inline src/qililab/qprogram/qblox_compiler.py:110]: bus_mapping argument missing in docstring

File: src/qililab/instruments/qblox/qblox_qrm.py (hunk 2 above)
Comment (fedonman) [reply src/qililab/instruments/qblox/qblox_qrm.py:149]: No, because the sequencer loop above is inside the acquisitions loop, so the command will be executed multiple times for each sequencer. Plus, if you do `delete_acquisition_data(sequencer=sequencer.identifier, all=True)` inside that loop you will delete all data before retrieving it with `store_scope_acquisition(sequencer=sequencer.identifier, name=acquisition)`. So you only want to delete the data after retrieving everything.

File: src/qililab/instruments/quantum_machines/quantum_machines_cluster.py [hunk 7]
```diff
@@ -282,34 +285,85 @@ def reset(self):
...
 
-        Returns:
-            bool: True if the parameter is set correctly, False otherwise
+        Raises:
+            NotImplementedError: Raised if not connected to Quantum Machines
+            ParameterNotFound: Raised if parameter does not exist
+        """
+        # TODO: This is a temporary solution. Implement again when QM API gets updated.
+        if not self._is_connected_to_qm:
+            raise NotImplementedError("You should be connected to Quantum Machines in order to change a parameter.")
```
[the remark this answers is not in the document]
Comment (fedonman) [reply src/qililab/instruments/quantum_machines/quantum_machines_cluster.py:321]: There is only one Quantum Machines instrument.

File: src/qililab/instruments/yokogawa/gs200.py (hunk 4 above)
Comment (fedonman) [reply src/qililab/instruments/yokogawa/gs200.py:202]: It's a bug we found with Juanjo on Yoko and decided to add it here for easiness since only Roccqet uses the Yoko. Will add an entry in changelog.

File: src/qililab/instruments/qblox/qblox_qrm.py (hunk 2 above)
Comment (fedonman) [reply src/qililab/instruments/qblox/qblox_qrm.py:149]: No, because the sequencer loop above is inside the acquisitions loop, so the command will be executed multiple times for each sequencer. Plus, if you do `delete_acquisition_data(sequencer=sequencer.identifier, all=True)` inside that loop you will delete all data before retrieving it with `store_scope_acquisition(sequencer=sequencer.identifier, name=acquisition)`. So you only want to delete the data after retrieving everything.

File: src/qililab/instruments/quantum_machines/quantum_machines_cluster.py [hunk 8]
```diff
@@ -282,34 +285,85 @@ def reset(self):
...
-        self._config = merge_dictionaries(dict(self._config), compilation_config)
-        # If we are already connected, reopen the connection with the new configuration
-        if self._is_connected_to_qm():
-            self._qm = self._qmm.open_qm(config=self._config, close_other_machines=True)
+        merged_configuration = merge_dictionaries(dict(self._config), configuration)
+        if self._config != merged_configuration:
+            self._config = cast(DictQuaConfig, merged_configuration)
+            # If we are already connected, reopen the connection with the new configuration
+            if self._is_connected_to_qm:
+                self._qm = self._qmm.open_qm(config=self._config, close_other_machines=True)
```
[the remark this answers is not in the document]
Comment (fedonman) [reply src/qililab/instruments/quantum_machines/quantum_machines_cluster.py:299]: If you are not connected then there is no need to re-open the connection. When change the `config` directory you have to open a new QuantumMachine instance. So, this checks if we have already opened a QuantumMachine and if yes, then we reopen with the new configuration.

File: src/qililab/platform/platform.py (hunk 5 above)
Comment (fedonman) [reply src/qililab/platform/platform.py]: Done. Also changed the filenames.

File: src/qililab/instruments/quantum_machines/quantum_machines_cluster.py (hunk 3 above)
Comment (fedonman) [reply src/qililab/instruments/quantum_machines/quantum_machines_cluster.py:324]: No, if `"RF_inputs" in self._config["elements"][bus]) == False` then you will just not update the `_config` dictionary. But you will not raise a `ParameterNotFound` error either. 

In general the implementation of `get_parameter` and `set_parameter` is super temporary until the QM API is improved (I had a meeting with Theo on that and they will try to make the changes soon)

This is why I don't want to mess with too many checks and exception raising. I assume that the ones who use it right now, know what they are doing. And also I assume we work only with Octave and not some external up/downconverter.

File: src/qililab/qprogram/qblox_compiler.py (hunk 6 above)
Comment (fedonman) [reply src/qililab/qprogram/qblox_compiler.py:110]: Done

File: src/qililab/instruments/quantum_machines/quantum_machines_cluster.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (fedonman) [reply src/qililab/instruments/quantum_machines/quantum_machines_cluster.py]: Changed it after all.

File: src/qililab/instruments/quantum_machines/quantum_machines_cluster.py (hunk 3 above)
Comment (fedonman) [reply src/qililab/instruments/quantum_machines/quantum_machines_cluster.py]: Eventually, I re-implemented the method to be more robust, and also save the changes to `settings` as well.

File: src/qililab/instruments/qblox/qblox_qrm.py (hunk 2 above)
Comment (visagim) [reply src/qililab/instruments/qblox/qblox_qrm.py:149]: fair enough, why not switch the order of loops then?

```python
        for sequencer in self.awg_sequencers:
            for acquisition in acquisitions:
                if sequencer.identifier in self.sequences:
                    self.device.get_acquisition_state(
                        sequencer=sequencer.identifier,
                        timeout=cast(AWGQbloxADCSequencer, sequencer).acquisition_timeout,
                    )
                    self.device.store_scope_acquisition(sequencer=sequencer.identifier, name=acquisition)
                    raw_measurement_data = self.device.get_acquisitions(sequencer=sequencer.identifier)[acquisition][
                        "acquisition"
                    ]
                    measurement_result = QbloxQProgramMeasurementResult(raw_measurement_data=raw_measurement_data)
                    results.append(measurement_result)

            self.device.delete_acquisition_data(sequencer=sequencer.identifier, all=True)
```
since `self.awg_sequencers` is always the same (it's a class attribute) switching the order of the loops should be the same

Comment (visagim) [reply src/qililab/instruments/qblox/qblox_qrm.py:149]: this is quite trivial though, if you prefer having the delete data later in a separate loop I don't think it's that different performance wise

File: src/qililab/instruments/quantum_machines/quantum_machines_cluster.py [hunk 9]
```diff
@@ -282,34 +285,112 @@ def reset(self):
...
+        # TODO: Change private QM API to public when implemented.
+        if not self._is_connected_to_qm:
+            raise NotImplementedError(f"You should be connected to {self.name} in order to change a parameter.")
+
+        element = next((element for element in self.settings.elements if element["bus"] == bus), None)
+        if element is None:
+            raise ValueError(f"Bus {bus} was not found in {self.name} settings.")
+
+        if parameter in [Parameter.LO_FREQUENCY, Parameter.GAIN]:
+            if "rf_inputs" not in element:
```
Comment (visagim) [inline src/qililab/instruments/quantum_machines/quantum_machines_cluster.py:328]: is `GAIN` controlled in the LO (i guess that's the octave) for QM? just curious

File: tests/instruments/quantum_machines/test_quantum_machines_cluster.py [hunk 10]
```diff
@@ -214,7 +250,158 @@ def test_simulate(
...
+        if parameter == Parameter.GAIN:
+            qmm_with_octave._qm.octave.set_rf_output_gain.assert_called_once()
+        if parameter == Parameter.IF:
+            qmm_with_octave._qm.set_intermediate_frequency.assert_called_once()
+
+    @pytest.mark.parametrize(
+        "bus, parameter, value",
+        [
+            ("drive_q0", Parameter.IF, 20e6),
+        ],
```
Comment (visagim) [inline tests/instruments/quantum_machines/test_quantum_machines_cluster.py:292]: this you leave this in parametrize because you expect to add more parameters to be set by the non-octave buses in the future?

**Review by visagim — CHANGES_REQUESTED** [review]
Cool! Just some minor changes. Once you confirm that this has been tested on hardware we can merge.

File: src/qililab/instruments/qblox/qblox_qrm.py (hunk 2 above)
Comment (fedonman) [reply src/qililab/instruments/qblox/qblox_qrm.py:149]: There is no performance drawback of having a linear loop, especially when it has maximum 6 elements 😅

File: src/qililab/instruments/quantum_machines/quantum_machines_cluster.py (hunk 9 above)
Comment (fedonman) [reply src/qililab/instruments/quantum_machines/quantum_machines_cluster.py:328]: Yes, it's this one: https://docs.quantum-machines.co/1.1.6/qm-qua-sdk/docs/API_references/qm_octave/#qm.octave.qm_octave.QmOctave.set_rf_output_gain

File: tests/instruments/quantum_machines/test_quantum_machines_cluster.py (hunk 10 above)
Comment (fedonman) [reply tests/instruments/quantum_machines/test_quantum_machines_cluster.py:292]: Exactly. We will certainly add more parameters in the (near) future, but these 3 are the most important right now.

File: src/qililab/qprogram/qprogram.py [hunk 11]
```diff
@@ -316,6 +335,7 @@ def set_gain(self, bus: str, gain: float):
         """
         operation = SetGain(bus=bus, gain=gain)
         self._active_block.append(operation)
+        self._buses.add(bus)
 
     def set_offset(self, bus: str, offset_path0: float, offset_path1: float):
```
Comment (jordivallsq) [inline src/qililab/qprogram/qprogram.py:340]: Shouldn't this function need also the `@requires_domain("gain", Domain.Voltage)`?

Comment (fedonman) [reply src/qililab/qprogram/qprogram.py:340]: Yes, you right. Nice catch. I will add it.

Comment (fedonman) [reply src/qililab/qprogram/qprogram.py:342]: Done

**Review by visagim — APPROVED** [review]
Go ahead
